Firmware handoff

raymo200915 opened the pull request:
Add supports for Transfer List into core, which are compliant to the Firmware Handoff
specification v0.9 (https://github.com/FirmwareHandoff/firmware_handoff/releases/tag/v0.9).

The motivation for this patch set is to unify the way of arguments handoff between all boot stages (BL2/BL31/BL32/BL33).
In current practice, when the Transfer List is enabled between BL2 and BL33 (See the related patch sets for TF-A below for reference),
we need this patch to allow the DTB nodes appended by OP-TEE can be reflected into the Transfer List memory and passed to BL33.

Reference:
1. Patch to introduce Transfer List into TF-A: https://review.trustedfirmware.org/c/TF-A/trusted-firmware-a/+/22215
2. Patch to enable Transfer List between the handoff from BL2 to BL33 on qemu: https://review.trustedfirmware.org/c/TF-A/trusted-firmware-a/+/22178

jenswi-linaro commented:
Please rebase to resolve the conflicts. We also need to decide if firmware handoff will be used by OP-TEE or not. A good use case is enough motivation to include this.

etienne-lms reviewed (commented):
Some generic coding style comments that could be addressed before you rebase.
Could you add a link the the Firmware Handoff Transfer List specification?

TL is too vague I think as acronym. If you want an acronym, maybe FHTL would be better (Firmware Handoff Transfer List, as its the name of the spec).

raymo200915 commented:
Updated the patch set with:
1. Fix to enhance the backwards compatibility for legacy handoff if no Transfer List is detected through the arguments.
2. Fix of coding style issues (False-positive spelling warnings are still remained).
3. Rebase to master.

raymo200915 commented:
> Please rebase to resolve the conflicts. We also need to decide if firmware handoff will be used by OP-TEE or not. A good use case is enough motivation to include this.

Rebased and add description on the motivation, links of specification and reference.

raymo200915 commented:
> [fixup] core: init dt when no DTB observed from TL

Fix in the commit '[[fixup] core: init dt when no DTB observed from TL](https://github.com/OP-TEE/optee_os/pull/6308/commits/6fb3c95a04cc178a08d0a9a88a5571e27947df0a)

raymo200915 commented:
> Some generic coding style comments that could be addressed before you rebase. Could you add a link the the Firmware Handoff Transfer List specification?
> 
> TL is too vague I think as acronym. If you want an acronym, maybe FHTL would be better (Firmware Handoff Transfer List, as its the name of the spec).

Fix in the commit [[fixup] core: init dt when no DTB observed from TL](https://github.com/OP-TEE/optee_os/pull/6308/commits/6fb3c95a04cc178a08d0a9a88a5571e27947df0a)

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,533 @@
+
+	if (!tl) {
+		return;
+	}
```
Fix in the commit [[fixup] core: init dt when no DTB observed from TL](https://github.com/OP-TEE/optee_os/pull/6308/commits/6fb3c95a04cc178a08d0a9a88a5571e27947df0a)

raymo200915 commented on core/include/kernel/transfer_list.h:
```diff
@@ -0,0 +1,116 @@
+
+// version of the register convention used.
+// Set to 1 for both AArch64 and AArch32 according to fw handoff spec v0.9
+#define REGISTER_CONVENTION_VERSION_MASK (1 << 24)
```
Fix in the commit [[fixup] core: init dt when no DTB observed from TL](https://github.com/OP-TEE/optee_os/pull/6308/commits/6fb3c95a04cc178a08d0a9a88a5571e27947df0a)

raymo200915 commented on core/include/mm/core_mmu.h:
```diff
@@ -137,6 +138,7 @@ static inline const char *teecore_memtype_name(enum teecore_memtypes type)
 		[MEM_AREA_IO_NSEC] = "IO_NSEC",
 		[MEM_AREA_IO_SEC] = "IO_SEC",
 		[MEM_AREA_EXT_DT] = "EXT_DT",
+		[MEM_AREA_TRANSFER_LIST] = "MEM_AREA_TRANSFER_LIST",
```
Fix in the commit [[fixup] core: init dt when no DTB observed from TL](https://github.com/OP-TEE/optee_os/pull/6308/commits/6fb3c95a04cc178a08d0a9a88a5571e27947df0a)

raymo200915 commented on core/include/mm/core_mmu.h:
```diff
@@ -104,6 +104,7 @@ enum teecore_memtypes {
 	MEM_AREA_IO_NSEC,
 	MEM_AREA_IO_SEC,
 	MEM_AREA_EXT_DT,
+	MEM_AREA_TRANSFER_LIST,
```
Fixed in [[fixup] core: fix coding style issues](https://github.com/OP-TEE/optee_os/pull/6308/commits/5a08df2a723b7f0f27eac709c26cff7b43f6c2a4)

raymo200915 commented on core/include/kernel/transfer_list.h:
```diff
@@ -0,0 +1,116 @@
+#define TRANSFER_LIST_VERSION		U(0x0001)
+
+// Init value of maximum alignment required by any TE data in the TL
+// specified as a power of two
```
Fixed in [[fixup] core: fix coding style issues](https://github.com/OP-TEE/optee_os/pull/6308/commits/5a08df2a723b7f0f27eac709c26cff7b43f6c2a4)

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,533 @@
+	DMSG("hdr_size   0x%x\n", tl->hdr_size);
+	DMSG("alignment  0x%x\n", tl->alignment);
+	DMSG("size       0x%x\n", tl->size);
+	DMSG("max_size   0x%x\n", tl->max_size);
```
Fixed in [[fixup] core: fix coding style issues](https://github.com/OP-TEE/optee_os/pull/6308/commits/5a08df2a723b7f0f27eac709c26cff7b43f6c2a4)

raymo200915 commented on mk/config.mk:
```diff
@@ -494,6 +494,11 @@ ifeq ($(CFG_MAP_EXT_DT_SECURE),y)
 $(call force,CFG_DT,y)
 endif
 
+CFG_TRANSFER_LIST ?= n
```
Fixed in [[fixup] core: fix coding style issues](https://github.com/OP-TEE/optee_os/pull/6308/commits/5a08df2a723b7f0f27eac709c26cff7b43f6c2a4)

raymo200915 commented:
All comments have been addressed.

jenswi-linaro commented:
Please squash the fixups.

raymo200915 commented:
Squash done.

jenswi-linaro commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,524 @@
+ * Compliant to 2.4.5 of Firmware handoff specification (v0.9)
+ * Return pointer to the created transfer list or NULL on error
+ ******************************************************************************/
+struct transfer_list_header *transfer_list_init(void *pa, size_t max_size)
```
It seems that `paddr_t pa` would be more suitable.

raymo200915 replied:
Fixed in the fixup patches.

jenswi-linaro commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,524 @@
+	if (!tl || !pa || max_size == 0)
+		return NULL;
+
+	align_mask = (1 << tl->alignment) - 1;
```
`BIT(tl->alignment) - 1.

raymo200915 replied:
It will be a bit confused to use 'BIT' - Here it means 'pow(2, tl->alignment)', but not the BIT3.

raymo200915 replied:
Replaced `1 << tl->alignment` with `GET_TL_ALIGNMENT(tl)`.

jenswi-linaro commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,524 @@
+	new_addr = ((paddr_t)pa & ~align_mask) + align_off;
+
+	if (new_addr < (paddr_t)pa)
+		new_addr += (1 << tl->alignment);
```
`BIT(tl->alignment)`

raymo200915 replied:
Same, it will be a bit confused to use 'BIT' - Here it means 'pow(2, tl->alignment)'.

raymo200915 replied:
Replaced `1 << tl->alignment` with `GET_TL_ALIGNMENT(tl)`.

jenswi-linaro commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,524 @@
+
+/*******************************************************************************
+ * Calculate the byte sum of a transfer list
+ * Return byte sum of the transfer list
```
Might be worth mentioning that this is modulo 256.

raymo200915 replied:
Fixed in the fixup patches.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,524 @@
+	if (!pa || max_size == 0)
+		return NULL;
+
+	if (!IS_ALIGNED((paddr_t)pa, 1 << TRANSFER_LIST_INIT_MAX_ALIGN) ||
```
I guess it will be a bit confused to use 'BIT' - Here it means 'pow(2, 3)'.

raymo200915 replied:
Added TL_ALIGNMENT macro to calculate the alignment from power of 2.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,524 @@
+ * Compliant to 2.4.6 of Firmware handoff specification (v0.9)
+ * Return true on success or false on error
+ ******************************************************************************/
+struct transfer_list_header *transfer_list_relocate
```
Currently, no. But according to the specification of the Firmware Handoff, the ability to relocating an existing TL should be reserved in all the boot stages.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,524 @@
+ * Compliant to 2.4.5 of Firmware handoff specification (v0.9)
+ * Return pointer to the created transfer list or NULL on error
+ ******************************************************************************/
+struct transfer_list_header *transfer_list_init(void *pa, size_t max_size)
```
Currently, no. But according to the specification of the Firmware Handoff, the ability to create(init) a TL should be reserved in all the boot stages.

raymo200915 commented on core/arch/arm/kernel/entry_a64.S:
```diff
@@ -326,8 +356,20 @@ clear_nex_bss:
+#if defined(CFG_TRANSFER_LIST)
+	/* check if the TL header valid */
+	mov	x0, x23
+	bl	transfer_list_check_header
```
I was trying to make it backward compatible - When the TL header is invalid then switch back to the old handoff way.
So, do you prefer a panic when we have a TL signature but with an invalid header? In this case we have less compatibility I guess.

jenswi-linaro replied:
The problem is that if a TL signature but an invalid TL header, there's nothing to fall back on. If the boot loader passes a TL it better not be corrupt.

raymo200915 replied:
Fair enough. I have updated a commit to panic system when the TL header is invalid.

raymo200915 commented on core/include/kernel/transfer_list.h:
```diff
@@ -0,0 +1,127 @@
+void transfer_list_dump(struct transfer_list_header *tl);
+struct transfer_list_header *transfer_list_init(void *pa, size_t max_size);
+
+struct transfer_list_header *transfer_list_relocate
```
Fixed in the fixup patches.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,524 @@
+ ******************************************************************************/
+struct transfer_list_header *transfer_list_init(void *pa, size_t max_size)
+{
+	struct transfer_list_header *tl;
```
Fixed in the fixup patches.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,524 @@
+					(struct transfer_list_header *tl,
+					 void *pa, size_t max_size)
+{
+	paddr_t new_addr, align_mask, align_off;
```
Fixed in the fixup patches.

raymo200915 commented on core/arch/arm/kernel/entry_a64.S:
```diff
@@ -160,8 +161,28 @@
+	cmp	x1, x23
+	b.eq	tl_sig_exist
+#endif
+	/* Transfer list signature does not exist, legacy mode
```
Fixed in the fixup patches.

raymo200915 commented on core/arch/arm/kernel/entry_a64.S:
```diff
@@ -184,7 +206,15 @@ FUNC _start , :
+	mov	x23, #0
+	b	sav_arg_done
+
+tl_sig_exist:	/* Transfer list signature exists */
```
Fixed in the fixup patches.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,524 @@
+ ******************************************************************************/
+struct transfer_list_header *transfer_list_relocate
+					(struct transfer_list_header *tl,
+					 void *pa, size_t max_size)
```
Fixed in the fixup patches.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,524 @@
+/*******************************************************************************
+ * Relocating a transfer list to a reserved memory region specified
+ * Compliant to 2.4.6 of Firmware handoff specification (v0.9)
+ * Return true on success or false on error
```
Fixed in the fixup patches.

raymo200915 commented on core/arch/arm/kernel/entry_a64.S:
```diff
@@ -382,6 +424,7 @@ clear_nex_bss:
 #endif
 	mov	x1, #-1
+	mov	x2, x23		/* TL address */
 	bl	boot_init_primary_early
```
Fixed in the fixup patches.

raymo200915 commented on core/arch/arm/kernel/entry_a64.S:
```diff
@@ -326,8 +356,20 @@ clear_nex_bss:
 #ifdef CFG_CORE_ASLR
 	mov	x0, x20		/* DT address */
+	mov	x1, x23		/* TL address */
 	bl	get_aslr_seed
```
Fixed in the fixup patches.

raymo200915 commented on core/arch/arm/kernel/entry_a64.S:
```diff
@@ -160,19 +161,40 @@
+	ldr     x20, =REGISTER_CONVENTION_VERSION_MASK
+	orr     x23, x19, x20
+	cmp	x1, x23
+	b.eq	tl_sig_exist
```
But then how can we do a fall back to support the "legacy" handoff if the TL signature does not exist?

raymo200915 replied:
Done.

raymo200915 commented on core/include/kernel/transfer_list.h:
```diff
@@ -0,0 +1,125 @@
+	 * traversed in order
+	 *
+	 * struct transfer_list_entry entries[];
+	 */
```
No, I think we don't need an empty TE array here. The commented block just want to tell what kind of data structure will be dynamically following the TL header in the memory.

raymo200915 commented on core/include/kernel/transfer_list.h:
```diff
@@ -0,0 +1,125 @@
+	uint16_t tag_id;
+	uint8_t reserved0;	/* place holder */
+	uint8_t hdr_size;
+	uint32_t data_size;
```
Same, I think we don't need a placeholder here. The commented block just want to tell what kind of data structure will be dynamically following the TE header in the memory.

raymo200915 commented on core/kernel/dt.c:
```diff
@@ -553,23 +554,36 @@ void init_external_dt(unsigned long phys_dt)
+	if (IS_ENABLED(CFG_TRANSFER_LIST))
+		/* get DTB from the mapped TL */
+		fdt = transfer_list_entry_data
+			(transfer_list_find(dt->tl, TL_TAG_FDT));
```
What is the motivation for this suggestion? And `transfer_list_find()` will return a `transfer_list_entry *` but not `transfer_list_header *`

etienne-lms replied:
Rational regarding the braces is that the inline comment + the instruction make the block a multi line block so prefer have braces:
```c
	if (IS_ENABLED(CFG_TRANSFER_LIST)) {
		/* get DTB from the mapped TL */
		fdt = transfer_list_entry_data
			(transfer_list_find(dt->tl, TL_TAG_FDT));
	}
```

Rational for an extra local variable is to prevent the line above break,
and comply with 80char/line max. Or could break line after the function opening parenthesis:
```c
	if (IS_ENABLED(CFG_TRANSFER_LIST)) {
		/* get DTB from the mapped TL */
		fdt = transfer_list_entry_data(
			transfer_list_find(dt->tl, TL_TAG_FDT));
	}
```

raymo200915 replied:
'make checkpatch' complains 'CHECK: Lines should not end with a '('' for your second suggestion.
I think it is fair enough to just remove the inline comment.

raymo200915 commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1253,13 +1254,30 @@ static void init_secondary_helper(unsigned long nsec_entry)
+
+		te = transfer_list_find(tl, TL_TAG_OPTEE_PAGABLE_PART);
+		if (IS_ENABLED(CFG_WITH_PAGER) && !pageable_part && te) {
+			pageable_part = get_le64(transfer_list_entry_data(te));
```
You are right. I think here we can check `pageable_part` first, this must be 0 if we are passing a valid TL.

raymo200915 commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1253,13 +1254,30 @@ static void init_secondary_helper(unsigned long nsec_entry)
+		te = transfer_list_find(tl, TL_TAG_OPTEE_PAGABLE_PART);
+		if (IS_ENABLED(CFG_WITH_PAGER) && !pageable_part && te) {
+			pageable_part = get_le64(transfer_list_entry_data(te));
+			transfer_list_rem(tl, te);
```
To remove or keep a TE after it is consumed is fully arbitrary at the moment (So far the spec does not mention).
I can keep this TE unless any changes in the spec.

raymo200915 commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1333,6 +1351,14 @@ unsigned long __weak get_aslr_seed(void *fdt)
+			EMSG("fdt does not match to the fdt entry of the TL");
+			goto err;
+		}
+	}
```
I can put all TL sanity check into `boot_save_transfer_list`

raymo200915 commented on core/kernel/dt.c:
```diff
@@ -614,12 +630,23 @@ static TEE_Result release_external_dt(void)
+			transfer_list_set_data_size
+				(tl, te, fdt_totalsize(external_dt.blob));
+
+			transfer_list_unmap_sync(tl); // unmap the TL
```
Done.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,528 @@
+		return TL_OPS_NON;
+
+	if (tl->signature != TRANSFER_LIST_SIGNATURE) {
+		EMSG("Bad transfer list signature %#"PRIx32"", tl->signature);
```
Done.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,528 @@
+		EMSG("Transfer list version is invalid");
+		return TL_OPS_NON;
+	} else if (tl->version == TRANSFER_LIST_VERSION) {
+		IMSG("Transfer list version is valid for all operations");
```
Done.

raymo200915 commented on core/arch/arm/kernel/entry_a32.S:
```diff
@@ -499,8 +532,14 @@ shadow_stack_access_ok:
 #endif
 
+#if defined(CFG_TRANSFER_LIST)
+	/* check if the TL header valid */
```
Done.

raymo200915 commented on core/arch/arm/kernel/entry_a64.S:
```diff
@@ -160,19 +161,40 @@
+	cmp	x1, x23
+	b.eq	tl_sig_exist
+#endif
+	/* No transfer list signature exist
```
Done.

raymo200915 commented on core/include/kernel/transfer_list.h:
```diff
@@ -0,0 +1,125 @@
+ */
+#define TRANSFER_LIST_INIT_MAX_ALIGN	U(3)
+
+/* alignment required by TE header start address, in bytes */
```
Done.

raymo200915 commented:
All comments have been addressed.

etienne-lms commented on core/kernel/dt.c:
```diff
@@ -614,12 +629,23 @@ static TEE_Result release_external_dt(void)
-	if (core_mmu_remove_mapping(MEM_AREA_EXT_DT, external_dt.blob,
-				    CFG_DTB_MAX_SIZE))
+	if (IS_ENABLED(CFG_TRANSFER_LIST)) {
+		if (tl) {
```
should be `if (IS_ENABLED(CFG_TRANSFER_LIST) && tl)`
so that when `CFG_TRANSFER_LIST=y` but the boot uses external FDT, we unmap from line 642.

raymo200915 replied:
Done.

etienne-lms commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,522 @@
+{
+	struct transfer_list_header *tl = NULL;
+
+	if (!pa || max_size == 0)
```
prefer `!pa || !max_size`

raymo200915 replied:
Done.

etienne-lms commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,522 @@
+	    !IS_ALIGNED(max_size, TL_ALIGNMENT(TRANSFER_LIST_INIT_MAX_ALIGN)) ||
+	    max_size < sizeof(*tl)) {
+		return NULL;
+	}
```
should remove the brace `{}`. Checkpatch will likely complain.

raymo200915 replied:
Fixed.

etienne-lms commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,522 @@
+	struct transfer_list_header *new_tl = NULL;
+	size_t new_max_size = 0;
+
+	if (!tl || !pa || max_size == 0)
```
`!max_size`

raymo200915 replied:
Done.

etienne-lms commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,522 @@
+				 TL_ALIGNMENT(tl->alignment));
+	new_max_size = max_size - (new_addr - pa);
+
+	/* the new space is not sufficient for the tl */
```
Nitpicking: prefer `/* The new space is not sufficient for the TL */`

raymo200915 replied:
Done.

etienne-lms commented on core/kernel/dt.c:
```diff
@@ -553,23 +554,35 @@ void init_external_dt(unsigned long phys_dt)
 		return;
 
-	if (!phys_dt) {
+	if (IS_ENABLED(CFG_TRANSFER_LIST))
```
There are dependencies `CFG_EXTERNAL_DT=y`. Should _mk/config.mk_ enforce that switch when `CFG_TRANSFER_LIST` is enabled?

raymo200915 replied:
Add enforcing `CFG_EXTERNAL_DT` when `CFG_TRANSFER_LIST` is defined.

jenswi-linaro commented:
The handling of boot arguments becomes even more complicated with this. I'm about to propose a PR to use common handling of boot arguments in C.

jenswi-linaro commented:
https://github.com/OP-TEE/optee_os/pull/6335

raymo200915 commented on core/arch/arm/kernel/entry_a32.S:
```diff
@@ -162,6 +163,24 @@ END_FUNC reset_vect_table
+	/*
+	 * FW Handoff register use:
+	 * entry r0: Zero
+	 * entry r1: TRANSFER_LIST_SIGNATURE | REGISTER_CONVENTION_VERSION_MASK
```
I think it should be noted here, as one of the entries.

raymo200915 commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1256,14 +1259,47 @@ void __weak boot_init_primary_early(unsigned long pageable_part,
 	unsigned long e = PADDR_INVALID;
+	struct transfer_list_entry *te = NULL;
+	struct transfer_list_header *tl = NULL;
+	struct dt_descriptor *dt = get_external_dt_desc();
```
I think the coding rules require to declare all local variables at the beginning of a function, am I right?

raymo200915 replied:
Fixed.

raymo200915 commented on core/include/kernel/transfer_list.h:
```diff
@@ -0,0 +1,139 @@
+	typeof(__pa) __new_addr = (__pa & ~__align_mask) + __align_off; \
+	\
+	__new_addr < __pa ? __new_addr + __a : __new_addr; \
+}))
```
I can add this, but just a bit confused since I followed the same styles in compile.h and util.h.

raymo200915 commented:
> The handling of boot arguments becomes even more complicated with this. I'm about to propose a PR to use common handling of boot arguments in C.

Just took a quick look at the PR, seems that all arguments are saved instead of passing into individual boot functions.
I can hold this patch after addressing remained comments and rebase all after your PR is merged.

raymo200915 commented on core/arch/arm/kernel/boot.c:
```diff
@@ -86,6 +87,8 @@ static void *manifest_dt __nex_bss;
 static uint32_t cntfrq;
 #endif
 
+static unsigned long pa_tl;
```
Done.

raymo200915 commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1256,14 +1259,47 @@ void __weak boot_init_primary_early(unsigned long pageable_part,
 #if !defined(CFG_WITH_ARM_TRUSTED_FW)
 	e = nsec_entry;
 #endif
+	if (IS_ENABLED(CFG_TRANSFER_LIST) && pa_tl && !pageable_part) {
```
Done.

raymo200915 commented:
All comments have been addressed.

jenswi-linaro commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,521 @@
+
+	do {
+		te = transfer_list_next(tl, te);
+	} while (te && (te->tag_id != tag_id || te->reserved0 != 0));
```
Why do we need to check `te->reserved0`?

raymo200915 replied:
Currently the spec reserves 3 bytes for `tag_id`, however all defined TEs are less than 2 bytes.
In order to simplify the API, I just use a ` uint16_t` for the TEs, but I still want to ensure the placeholder `reserved0` is 0 to avoid protential undefined TE tags which have the same lower bytes as the defined ones but a different upper byte.
But I think here I should replace `te->reserved0 != 0` with `te->reserved0`.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,521 @@
+	tl_ev = (vaddr_t)tl + tl->size;
+	ev = tl_ev + sizeof(struct transfer_list_entry);
+
+	if (!IS_ALIGNED(ev, 1 << alignment)) {
```
In fact, the term `alignment` in the spec is defined as 'powers of two' (See the spec in page 5).
Here I am trying to align with the spec to define the arguments (At least, at API level). I will add some more comments here.
And I think I can chage `1 << alignment` to `TL_ALIGNMENT_FROM_ORDER(alignment)`.

raymo200915 commented on core/include/kernel/transfer_list.h:
```diff
@@ -0,0 +1,140 @@
+#include <types_ext.h>
+
+/* Align the PA for a TL to fulfill the alignment requirement */
+#define TL_ADDR_ALIGN(pa, tl, a) (__extension__({ \
```
Moved to a static helper function.

raymo200915 commented on core/include/kernel/transfer_list.h:
```diff
@@ -0,0 +1,140 @@
+	}))
+
+/* Get alignment from a value specified as power of two */
+#define TL_ALIGNMENT(a) (1 << (a))
```
Fixed.

raymo200915 commented:
All comments have been addressed.
If there no upcoming comments I will squash all `[review]` commits and add `Reviewed-by` tags.

etienne-lms reviewed (commented):
I fine you squash the fixup commits.
Some minor late comments.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,535 @@
+		DMSG("tag_id     %#"PRIx16, te->tag_id);
+		DMSG("hdr_size   %#"PRIx8, te->hdr_size);
+		DMSG("data_size  %#"PRIx32, te->data_size);
+		DMSG("data_addr  %#"PRIxVA,
```
Compiler prompts a `Wformat` error if without casting here.
Thus I changed it to:
```
		DMSG("data_addr  %#"PRIxVA,
		     (vaddr_t)transfer_list_entry_data(te));
```

raymo200915 commented on core/kernel/dt.c:
```diff
@@ -614,12 +629,20 @@ static TEE_Result release_external_dt(void)
+		assert(external_dt.blob ==
+			transfer_list_entry_data(te));
+		transfer_list_set_data_size
+			(tl, te, fdt_totalsize(external_dt.blob));
```
Fixed.

raymo200915 commented on core/kernel/dt.c:
```diff
@@ -614,12 +629,20 @@ static TEE_Result release_external_dt(void)
+	if (IS_ENABLED(CFG_TRANSFER_LIST) && tl) {
+		te = transfer_list_find(tl, TL_TAG_FDT);
+		assert(external_dt.blob ==
+			transfer_list_entry_data(te));
```
Fixed.

etienne-lms commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1256,14 +1259,55 @@ void __weak boot_init_primary_early(unsigned long pageable_part,
+	struct transfer_list_header *tl = (void *)transfer_list;
+	struct transfer_list_entry *te = NULL;
+
+	if (!IS_ALIGNED(transfer_list, TL_ALIGNMENT_FROM_ORDER(tl->alignment)))
```
sorry I missed this one. This line checks tl alignment. We can't get `tl->aligment` if `tl` is not aligned.
I think this needs a 2 step test:
```c
	if (!IS_ALIGNED_WITH(transfer_list, struct transfer_list_header) ||
	    !IS_ALIGNED(transfer_list, TL_ALIGNMENT_FROM_ORDER(tl->alignment)))
		panic("Transfer list base address is not aligned");

raymo200915 replied:
Yes, you are right, I missed this.
Fixed.

etienne-lms commented:
@raymo200915, actually no review tag were posted yet for the 3 commits. We only posted comments. 
Please update mine with `Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for the 3 commits.

etienne-lms commented:
Checkpatch report are all false positive, all about missspelled `te`:
```
WARNING: 'te' may be misspelled - perhaps 'the'?
#55: FILE: core/arch/arm/kernel/boot.c:1262:
+	struct transfer_list_entry *te = NULL;
 	                            ^^

raymo200915 commented:
> @raymo200915, actually no review tag were posted yet for the 3 commits. We only posted comments. Please update mine with `Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for the 3 commits. (edited: my A-b tag assuming [#6308 (comment)](https://github.com/OP-TEE/optee_os/pull/6308#discussion_r1357085243) is irrelevant or fixed)

Done. Sorry, I misunderstand the purpose of `Reviewed-by` tag.

raymo200915 commented:
All comments have been addressed.

etienne-lms reviewed (changes requested):
@raymo200915, could you remove also Jens' A-b tag. @jenswi-linaro didn't explicitily posted it.
@jenswi-linaro, @jforissier, could you enable CI for this P-R?

Despite my A-b tag, I think this change should come after #6335 which is almost ready to be merged. @raymo200915, would you mind giving a feedback on #6335 and if you agree, rebase your series on top of it?

raymo200915 commented:
> @raymo200915, could you remove also Jens' A-b tag. @jenswi-linaro didn't explicitily posted it. @jenswi-linaro, @jforissier, could you enable CI for this P-R?
> 
> Despite my A-b tag, I think this change should come after #6335 which is almost ready to be merged. @raymo200915, would you mind giving a feedback on #6335 and if you agree, rebase your series on top of it?

Yes, sure. I talked with @jenswi-linaro and will rebase this PR after [#6335](https://github.com/OP-TEE/optee_os/pull/6335) is merged. I will amend the tag when I rebase.
Now just on hold this PR for a while until rebasing.

jforissier commented:
@raymo200915

1. In "core: add memory area for transfer list": `s/handovered/handed over/`
2. In "core: add transfer list api": `s/api/API`, and I would like a `Link:` tag to the Firmware Handover spec.
3. Could you please rebase onto master?

raymo200915 commented:
> @raymo200915
> 
> 1. In "core: add memory area for transfer list": `s/handovered/handed over/`
> 2. In "core: add transfer list api": `s/api/API`, and I would like a `Link:` tag to the Firmware Handover spec.
> 3. Could you please rebase onto master?

Hi @jforissier I have rebased on top of master and fixed the commit messages as what you suggested.

raymo200915 commented:
Hi @jenswi-linaro @etienne-lms I have rebased on Jens' PR https://github.com/OP-TEE/optee_os/pull/6335, please review.

jenswi-linaro commented on core/include/kernel/dt.h:
```diff
@@ -83,6 +83,7 @@ struct dt_descriptor {
 #ifdef _CFG_USE_DTB_OVERLAY
 	int frag_id;
 #endif
+	struct transfer_list_header *tl;
```
I would prefer if we didn't add this. Is it possible to do without it?

raymo200915 replied:
Fixed.

raymo200915 commented:
All checkpatch warnings are false positive as below:
`WARNING: 'te' may be misspelled - perhaps 'the'?` or
`WARNING: 'TE' may be misspelled - perhaps 'THE'?`

jenswi-linaro commented on core/kernel/dt.c:
```diff
@@ -553,23 +554,35 @@ void init_external_dt(unsigned long phys_dt)
 
-	if (!phys_dt) {
+	if (IS_ENABLED(CFG_TRANSFER_LIST))
+		fdt = transfer_list_entry_data(transfer_list_find(dt->tl,
```
I'd rather add another argument to `init_external_dt()` so it becomes:
``` C
void init_external_dt(void *fdt, unsigned long phys_dt)
```
That way we don't need to involve transfer lists in this function, instead `fdt` is non-null we use it, else we map the `phys_dt` and use that address.

raymo200915 replied:
I realize that we already set `boot_arg_fdt` as same as `transfer_list_entry_data(transfer_list_find(dt->tl, TL_TAG_FDT))` via function `boot_save_transfer_list()`, I think we can deprecate this few lines and don't need to add a new argument, too.

raymo200915 replied:
Ah, no, here I still need to pass an argument to indicate if the DTB is already mapped and bypass the re-map. I will add an arg `void *mapped_dt`.

raymo200915 replied:
Found a way to check the mapping type of the PA to know whether it is mapped or not, now we don't need to add additional argument to `init_external_dt()`

jenswi-linaro commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1448,6 +1492,15 @@ void __weak boot_save_args(unsigned long a0, unsigned long a1 __unused,
+	 * a3(x3/r3)	- Transfer list base address
+	 * a4(x4/r4)	- Not used
+	 *
+	 * Scenario B: Legacy Handoff arguments
```
These are not legacy arguments. I'm not sure that Transfer List will be used with FF-A so it may give the wrong impression to list Transfer Lists as the primary scenario.

When `CFG_TRANSFER_LIST` is defined, it undeniably has precedence, but I don't think that will be defined for `CFG_CORE_SEL2_SPMC=y || CFG_CORE_EL3_SPMC=y`.

raymo200915 replied:
I think we will use TL with FF-A, please refer to the discussion in the link: https://github.com/FirmwareHandoff/firmware_handoff/pull/19
We plan to add FF-A TEs e.g. `XFERLIST_SPMC_MANIFEST`.

But at the moment we don't support it yet.
Shall we still keep the term `Legacy Handoff arguments` here but add more comments to state that currently FF-A is not yet supported? Because I guess finally all 'standalone' arguments should become 'legacy'.

jenswi-linaro replied:
Let's call these default arguments since that's what they are when TL isn't enabled or used.

raymo200915 replied:
Sounds good.

jenswi-linaro commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1252,15 +1256,55 @@ void __weak boot_init_primary_early(void)
+static void boot_save_transfer_list(unsigned long transfer_list __maybe_unused,
+				    unsigned long fdt __maybe_unused)
+{
+#if defined(CFG_TRANSFER_LIST)
```
Why is this ifdefed?

raymo200915 replied:
Fixed.

raymo200915 replied:
My bad, fixed.

raymo200915 replied:
I need a stub function of `boot_save_transfer_list()` here for clang build without `CFG_TRANSFER_LIST`, instead I should remove `if (IS_ENABLED(CFG_TRANSFER_LIST)`.

jenswi-linaro replied:
I would prefer having stubbed static inline functions in `transfer_list.h` when `CFG_TRANSFER_LIST` isn't defined, like for instance for `is_sp_ctx()` in `secure_partition.h`.

raymo200915 replied:
Done.

jenswi-linaro commented on core/arch/arm/kernel/boot.c:
```diff
@@ -87,6 +89,8 @@ static unsigned long boot_arg_pageable_part __nex_bss;
 static uint32_t cntfrq;
 #endif
 
+static unsigned long transfer_list_pa;
```
Please move this just below `boot_arg_pageable_part`. We should use the `__nex_bss` attribute too.

raymo200915 replied:
Fixed.

jenswi-linaro commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1252,15 +1256,55 @@ void __weak boot_init_primary_early(void)
+		tl = transfer_list_map(transfer_list_pa);
+		if (!tl)
+			panic("Failed to map transfer list");
+		dt->tl = tl; /* save the mapped TL */
```
I'd rather store this in a global `static struct transfer_list_header *boot_tl` or such.

raymo200915 replied:
Fixed.

jenswi-linaro commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1448,6 +1492,15 @@ void __weak boot_save_args(unsigned long a0, unsigned long a1 __unused,
 {
```
All of a1-a3 above are used now.

raymo200915 replied:
Fixed.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,535 @@
+#include <string.h>
+#include <util.h>
+
+static paddr_t get_align_base_addr(paddr_t pa,
```
This helper function is to adapt an adress for relocating the TL, the address must be aligned to the maximum alignment required by its TEs. It follows the spec v0.9 2.4.6.
I can add a few comments for this function.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,535 @@
+				   struct transfer_list_header *tl)
+{
+	paddr_t align_mask = TL_ALIGNMENT_FROM_ORDER(tl->alignment) - 1;
+	paddr_t align_off = virt_to_phys(tl) & align_mask;
```
Yes. You are right. I guess we can ignore the situation when the alignment is bigger than a page.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,535 @@
+		 */
+		new_tl_ev = ROUNDUP(ev, TL_ALIGNMENT_FROM_ORDER(alignment)) -
+			    sizeof(struct transfer_list_entry);
+		dummy_te_data_sz =
```
I don't think this will happen, as `TRANSFER_LIST_GRANULE` is defined as the same value as `sizeof(struct transfer_list_entry)`.
I can add an assert here.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,535 @@
+
+	do {
+		te = transfer_list_next(tl, te);
+	} while (te && (te->tag_id != tag_id || te->reserved0));
```
hmmm, this loop will stop when `te->tag_id == tag_id && !te->reserved0`, otherwise `NULL` when not found, am I right?

jenswi-linaro replied:
If it can't be anything but 0 there's no point in checking if it really is. My point is that if for some reason it isn't 0 we'll return a `te` with an unexpected tag.

raymo200915 replied:
OK. Sounds good. I can ignore checking `reserved0` here.

raymo200915 commented on core/include/kernel/transfer_list.h:
```diff
@@ -0,0 +1,126 @@
+	 * Commented out element used to visualize dynamic part of the
+	 * data structure.
+	 *
+	 * Note that padding is added at the end of @data to make to reach
```
Fixed.

raymo200915 commented on core/arch/arm/kernel/entry_a32.S:
```diff
@@ -445,10 +445,15 @@ shadow_stack_access_ok:
 	mov	r1, r5
-	mov	r2, r6
 	mov	r3, r7
+#if defined(CFG_TRANSFER_LIST)
```
Fixed.

raymo200915 commented:
All comments have been addressed.

jenswi-linaro commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1512,3 +1581,38 @@ void __weak boot_save_args(unsigned long a0, unsigned long a1 __unused,
+	if (!mapped_tl)
+		return TEE_SUCCESS;
+
+	if (dt) {
```
Please add a comment describing that `release_external_dt()` will do nothing if the external DT is located in the memory area of a transfer list since it uses a different memtype.

raymo200915 replied:
I added new comments in `release_external_dt()`, `init_external_dt()` and `release_transfer_list()` to increase the readibility.

jenswi-linaro commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1441,13 +1483,24 @@ static void get_sec_mem_from_manifest(void *fdt, paddr_t *base, size_t *size)
+	 *
+	 * Scenario A: FW Handoff via Transfer List
+	 * Note: FFA is not yet supported with Transfer List
+	 * a0	- DTB address or 0
```
(AArch64)

jenswi-linaro commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1441,13 +1483,24 @@ static void get_sec_mem_from_manifest(void *fdt, paddr_t *base, size_t *size)
+	 * a0	- DTB address or 0
+	 *	- must be 0 (Armv7)
+	 * a1	- TRANSFER_LIST_SIGNATURE | REG_CONVENTION_VER_MASK
+	 * a2	- must be 0
```
(AArch64)

raymo200915 replied:
Fixed.

jenswi-linaro commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1441,13 +1483,24 @@ static void get_sec_mem_from_manifest(void *fdt, paddr_t *base, size_t *size)
+	 * Scenario A: FW Handoff via Transfer List
+	 * Note: FFA is not yet supported with Transfer List
+	 * a0	- DTB address or 0
+	 *	- must be 0 (Armv7)
```
Armv7 or AArch32?

raymo200915 replied:
Fixed.

jenswi-linaro commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1441,13 +1483,24 @@ static void get_sec_mem_from_manifest(void *fdt, paddr_t *base, size_t *size)
 	 * Register use:
+	 *
+	 * Scenario A: FW Handoff via Transfer List
+	 * Note: FFA is not yet supported with Transfer List
```
FF-A

raymo200915 replied:
Fixed.

raymo200915 commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1252,15 +1257,52 @@ void __weak boot_init_primary_early(void)
+	struct transfer_list_header *tl = (void *)transfer_list;
+	struct transfer_list_entry *te = NULL;
+
+	if (!IS_ENABLED(CFG_TRANSFER_LIST))
```
Fixed.

raymo200915 commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1252,15 +1257,52 @@ void __weak boot_init_primary_early(void)
+	}
+
+	if (IS_ENABLED(CFG_WITH_PAGER)) {
+		pageable_part = IS_ENABLED(CFG_TRANSFER_LIST) && te ?
```
Fixed

raymo200915 commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1441,13 +1483,24 @@ static void get_sec_mem_from_manifest(void *fdt, paddr_t *base, size_t *size)
 	/*
 	 * Register use:
+	 *
+	 * Scenario A: FW Handoff via Transfer List
```
Done.

raymo200915 commented on core/arch/arm/kernel/boot.c:
```diff
@@ -82,11 +84,14 @@ static void *manifest_dt __nex_bss;
 static unsigned long boot_arg_fdt __nex_bss;
 static unsigned long boot_arg_nsec_entry __nex_bss;
 static unsigned long boot_arg_pageable_part __nex_bss;
+static unsigned long transfer_list_pa __nex_bss;
```
Fixed.

raymo200915 commented on core/arch/arm/kernel/boot.c:
```diff
@@ -82,11 +84,14 @@ static void *manifest_dt __nex_bss;
 static uint32_t cntfrq;
 #endif
 
+static struct transfer_list_header *mapped_tl;
```
Fixed.

raymo200915 commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1473,6 +1526,22 @@ void __weak boot_save_args(unsigned long a0, unsigned long a1 __unused,
 
+	if (IS_ENABLED(CFG_TRANSFER_LIST) &&
+	    a1 == (TRANSFER_LIST_SIGNATURE | REG_CONVENTION_VER_MASK)) {
+#if defined(CFG_ARM64_core)
```
Fixed.

raymo200915 commented:
All comments have been addressed.

jenswi-linaro commented:
The build errors are best fixed with stubbed static inline functions in transfer_list.h provided when configured without transfer list.

etienne-lms commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1471,8 +1515,30 @@ void __weak boot_save_args(unsigned long a0, unsigned long a1 __unused,
+	 * Note: FF-A is not yet supported with Transfer List
+	 * a0	- DTB address or 0 (AArch64)
+	 *	- must be 0 (AArch32)
+	 * a1	- TRANSFER_LIST_SIGNATURE | REG_CONVENTION_VER_MASK
```
Regarding Armv7 platforms, this assumes that when `CFG_TRANSFER_LIST` is enabled, value 0x016ed0ff is a forbidden value for "Armv7 standard bootarg # 1" (line 1502). From mainline Linux kernel, it seems such value is not used (linux [Documentation/arch/arm/booting.rst](https://elixir.bootlin.com/linux/v6.5/source/Documentation/arch/arm/booting.rst#L63) + [arch/arm/tools/mach-types](https://elixir.bootlin.com/linux/v6.5/source/arch/arm/tools/mach-types)), but is it enough?

Regarding Armv8+ platforms, this assumes TF-A BL31 loads a1 with a value that always differs from 0x016ed0ff. I think TF-A/BL31 loads 0 in a1 but this should be cross-checked.

raymo200915 replied:
Please refer to the related TF-A changes with the topics `rm/handoff` and `tl_bl31_bl32`.

etienne-lms commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1471,8 +1515,30 @@ void __weak boot_save_args(unsigned long a0, unsigned long a1 __unused,
+	 * a2	- must be 0 (AArch64)
+	 *	- DTB address or 0 (AArch32)
+	 * a3	- Transfer list base address
+	 * a4	- Not used
```
When `CFG_WITH_ARM_TRUSTED_FW=n` (Armv7), a4 hold Non-secure entry address. Should this be described here or is does firmware handoff transfer list supported only on Armv8+?

raymo200915 replied:
The TL is supposed to support both v7 and v8. I believe we need to define a new TE for non-secure entry later.
Currently we can only support the defined TEs described in in the spec.
I will add one more comment here to note that currently non-secure entry is not yet supported.

etienne-lms commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1512,3 +1578,41 @@ void __weak boot_save_args(unsigned long a0, unsigned long a1 __unused,
+	return TEE_SUCCESS;
+}
+
+boot_final(release_transfer_list);
```
`release_external_dt()` from _core/kernel/dt.c_ also packs and unmaps the external DTB and is also registered as a `boot_final` initcall handler. There may be conflict between the 2 depending on which is called first.

raymo200915 replied:
There is no issue about that, please check the latest commit, `release_external_dt` only do packing and un-mapping when the DTB memory is mapped in `MEM_AREA_EXT_DT`, when it find out that is mapped in other areas, it skips both packing and un-mapping so there will be no conflicts when `CFG_TRANSFER_LIST` is enabled.

etienne-lms commented on core/include/kernel/transfer_list.h:
```diff
@@ -0,0 +1,126 @@
+#define REG_CONVENTION_VER_MASK BIT(24)
+
+/* Transfer list operation codes */
+#define TL_OPS_NON	U(0) /* invalid for any operation */
```
`TL_OPS_NONE` ? (more explicit)

raymo200915 replied:
Fixed.

etienne-lms commented on mk/config.mk:
```diff
@@ -495,6 +495,15 @@ ifeq ($(CFG_MAP_EXT_DT_SECURE),y)
+CFG_TRANSFER_LIST ?= n
+ifeq ($(CFG_TRANSFER_LIST),y)
+$(call force,CFG_DT,y)
+$(call force,CFG_EXTERNAL_DT,y)
```
Transfer list expects the TL is located in secure memory. Should this also enable `CFG_MAP_EXT_DT_SECURE`, for consistency?

raymo200915 replied:
hmmm, looks reasonable, as `MEM_AREA_TRANSFER_LIST` is mapped to secure. `external_dt.blob` will be pointed to secure when `CFG_TRANSFER_LIST` is enabled. I will add this flag.

etienne-lms commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,542 @@
+}
+
+/*******************************************************************************
+ * Enumerate the next transfer entry
```
It would be nice that argument `last` is also described. I guess here last refers to the last parsed TE, from where we want to find the next TE.

raymo200915 replied:
I realize that 'last' might be not very accurate, I should use 'current' instead - We want to pass in the 'current' TE and then find its 'next'.
I will update it with the comment.

etienne-lms commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,542 @@
+/*
+ * Copyright (c) 2023, Linaro Limited
+ */
+
```
Could there be a link to the spec as an inline comment? It's currently only found in the commit message.

raymo200915 replied:
I will add the link to the spec's Github project https://github.com/FirmwareHandoff/firmware_handoff
Then we can have a consitent link and avoid to update the link each time to different spec release.

etienne-lms commented on core/include/kernel/transfer_list.h:
```diff
@@ -0,0 +1,126 @@
+
+struct transfer_list_entry {
+	uint16_t tag_id;
+	uint8_t reserved0;	/* place holder */
```
`/* place holder for tag ID 3rd byte (MSB) */`?

raymo200915 replied:
Fixed.

etienne-lms commented on core/include/kernel/transfer_list.h:
```diff
@@ -0,0 +1,126 @@
+#ifndef __KERNEL_TRANSFER_LIST_H
+#define __KERNEL_TRANSFER_LIST_H
+
+#define TRANSFER_LIST_SIGNATURE		U(0x006ed0ff)
```
From spec v0.9, it seems the expected signature value is 0x0040b10b.

raymo200915 replied:
Did you get the spec from this link? https://github.com/FirmwareHandoff/firmware_handoff/releases/tag/v0.9
Please see the Table1, "signature | 0x4 | 0x0 | The value of signature must be 0x6e_d0ff“.

raymo200915 commented on core/kernel/dt.c:
```diff
@@ -600,13 +605,18 @@ void *get_external_dt(void)
 
+	pa_dt = virt_to_phys(external_dt.blob);
+	if (core_mmu_get_type_by_pa(pa_dt) != MEM_AREA_EXT_DT)
+		return TEE_SUCCESS;
```
I don't think it is good to add a check of `CFG_TRANSFER_LIST` here, we need to decouple transfer list stuffs from the kernel/dt.c which is purely for setup the DTB.
kernel/dt.c should only handle the mapping/un-mapping in the area `MEM_AREA_EXT_DT`, if the memory is already mapped/un-mapped (Not only limited to transfer list), it should skip the related operations but just use the mapped address - I think this is reasonable.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,542 @@
+/*******************************************************************************
+ * Relocating a transfer list to a reserved memory region specified
+ * Compliant to 2.4.6 of Firmware handoff specification (v0.9)
+ * Return pointer to the relocated transfer list or NULL on error
```
There is no standard patterns for a function header now in the OP-TEE project, so I will just add parameter description for the major APIs using `@arg:` style.

raymo200915 commented:
> The build errors are best fixed with stubbed static inline functions in transfer_list.h provided when configured without transfer list.

I added one inline function `boot_save_transfer_list()` to minimize the changes needed.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,542 @@
+}
+
+/*******************************************************************************
+ * Relocating a transfer list to a reserved memory region specified
```
Done and update other function headers as well.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,542 @@
+
+/*******************************************************************************
+ * Relocating a transfer list to a reserved memory region specified
+ * Compliant to 2.4.6 of Firmware handoff specification (v0.9)
```
Done and updated other function headers as well.

raymo200915 commented:
All comments have been addressed.

etienne-lms commented on core/kernel/dt.c:
```diff
@@ -566,11 +566,17 @@ void init_external_dt(unsigned long phys_dt)
+		if (!dt->blob)
+			panic("Failed to map external DTB");
+	} else {
+		/* Save the DTB if it is already mapped in one memory area */
```
Rather "Save DTB **address**...". Suggestion to fit in 80char/line:
`		/* Get the DTB address if already mapped in a memory area */`

Same comment for line 571 above. It's not the DTB that is saved but its location. I would suggestion to simply remove the 3 last words:
`/* Map the DTB if it is not yet mapped */`

raymo200915 replied:
Fixed.

etienne-lms reviewed (commented):
I'm fine you squash the fixup commits.

raymo200915 commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1252,15 +1256,62 @@ void __weak boot_init_primary_early(void)
+		panic("Incorrect transfer list register convention");
+
+	if (!IS_ALIGNED_WITH_TYPE(transfer_list,
+				  struct transfer_list_header) ||
```
I agree, but that will exceed 80 characters per line. I have to keep it as-is.

etienne-lms replied:
it's exactly 80 chars, so it fits.

raymo200915 replied:
Ah, my bad. Now it is fixed. @etienne-lms

raymo200915 commented on core/include/kernel/transfer_list.h:
```diff
@@ -0,0 +1,126 @@
+#ifndef __KERNEL_TRANSFER_LIST_H
+#define __KERNEL_TRANSFER_LIST_H
+
+#define TRANSFER_LIST_SIGNATURE		U(0x006ed0ff)
```
I don't believe the signature should be changed after we have a stable version of the spec for backward compatibilty.
Let's keep this as-is and keep an eye to see next update of the spec.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,577 @@
+ * @tl: Pointer to the transfer list.
+ * @cur: Pointer to the current transfer entry where we want to search for the
+ *       next one.
+ * Return pointer to the next transfer entry or NULL on error.
```
As this is a helper function to enumerate the entries, I think the caller is only interested in whether can find a next entry or not.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,577 @@
+	new_tl->max_size = new_max_size;
+
+	transfer_list_update_checksum(new_tl);
+	transfer_list_unmap_sync(tl);
```
Yes, agree.

raymo200915 commented on core/include/kernel/transfer_list.h:
```diff
@@ -0,0 +1,126 @@
+#include <types_ext.h>
+
+/* Get alignment from a value specified as power of two */
+#define TL_ALIGNMENT_FROM_ORDER(a) BIT((a))
```
Done.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,577 @@
+	te->tag_id = tag_id;
+	te->reserved0 = 0;
+	te->hdr_size = sizeof(*te);
+	te->data_size = data_size;
```
Done.

raymo200915 commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,577 @@
+ * transfer list.
+ * @tl: Pointer to the transfer list.
+ * @tag_id: Tag ID to match a transfer entry.
+ * Return pointer to the found transfer entry or NULL on error.
```
Fixed.

raymo200915 commented:
All comments have been addressed.
Commits are squashed and pushed.

jenswi-linaro commented on core/kernel/dt.c:
```diff
@@ -566,11 +566,17 @@ void init_external_dt(unsigned long phys_dt)
+			panic("Failed to map external DTB");
+	} else {
+		/* Get the DTB address if already mapped in a memory area */
+		dt->blob = phys_to_virt(phys_dt, mtype, CFG_DTB_MAX_SIZE);
```
This can also fail if the mapped memory area is too small.

raymo200915 replied:
Yes, it should be panic if the mapping failed. Fixed.

jenswi-linaro commented:
For "core: add memory area for transfer list" please apply:
Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>

raymo200915 commented:
> For "core: add memory area for transfer list" please apply: `Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

Done.

raymo200915 commented:
All comments have been addressed.

jenswi-linaro commented:
For "core: add support for transfer list" please apply:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

For "core: add transfer list API" please apply:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

raymo200915 commented:
> For "core: add support for transfer list" please apply: `Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`
> 
> For "core: add transfer list API" please apply: `Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

Thanks @jenswi-linaro , tags added.

jforissier commented:
- There is a link error in "CI / make (multi-platform)"
- How about renaming `te` to `e` to please checkpatch? `e` for entry is as good as `te` for transfer list entry IMO.
- The Hafnium CI error is probably unrelated, I'll look into it

etienne-lms commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,580 @@
+		.data_size = data_size,
+	};
+
+	tl->size += ev - tl_ev;
```
I think `tl->size` should be updated only once `data` processing below succeeds

raymo200915 replied:
No. This function allows passing in NULL TE data pointer.
In this case, we still create the TE with the required size but without copying the data. The data region is just for placeholder.
The scenario for this is to add a 'gap' with an 'empty' tag to fulfill the alignment requirement, see the caller function `transfer_list_add_with_align()` for reference.

etienne-lms reviewed (commented):
comment for commit "core: add transfer list API".

etienne-lms commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,580 @@
+		/*
+		 * TE data address is not aligned to the new alignment
+		 * fill the gap with an empty TE as a placeholder before
+		 * adding the desire TE
```
s/desire/desired/

Few punctuation are missing + s/fil/Fill/:
```
 * TE data address is not aligned to the new alignment.
 * Fill the gap with an empty TE as a placeholder before
 * adding the desired TE.

raymo200915 replied:
Fixed.

etienne-lms commented on core/kernel/transfer_list.c:
```diff
@@ -0,0 +1,580 @@
+void transfer_list_dump(struct transfer_list_header *tl)
+{
+	struct transfer_list_entry *te = NULL;
+	int i = 0;
```
`__maybe_unused`

Could you build TL support with `CFG_TEE_CORE_LOG_LEVEL=0 CFG_TRANSFER_LIST=y CFG_WERROR=y`? I think it would have triggered an error here.

raymo200915 replied:
Fixed.

raymo200915 commented:
> * There is a link error in "CI / make (multi-platform)"
> * How about renaming `te` to `e` to please checkpatch? `e` for entry is as good as `te` for transfer list entry IMO.
> * The Hafnium CI error is probably unrelated, I'll look into it

Fixed the clang build error in multi-platform.
All `te` are renamed to `tl_e` for 'transfer list entry', no more checkpatch complains.

raymo200915 commented:
Minor fixes applied:
Adding a stub function "boot_save_transfer_list()" to fix the CI clang build error.
Optimize function `transfer_list_add()`.
Clean up checkpatch false positive warnings.

raymo200915 commented:
Fixed another CI multi-platform build error.

etienne-lms commented:
@raymo200915, it's hard to review again the whole changes after your latest updates since we don't really know which piece of code you updated. This is where appending fixup commits on top of an series of the patches we have reviewed (acked or not acked, not matter)  is quite handy for the dear reviewers.
Nevertheless i'll have another review round...

@jforissier, could you enable CI tests for this P-R?
@raymo200915, it would be nice the CI includes a build test with `CFG_TRANSFER_LIST=y`. You can archive that by patching file _.github/workflows/ci.yaml_, see that file Git history for examples.

raymo200915 commented:
> @raymo200915, it's hard to review again the whole changes after your latest updates since we don't really know which piece of code you updated. This is where appending fixup commits on top of an series of the patches we have reviewed (acked or not acked, not matter) is quite handy for the dear reviewers. Nevertheless i'll have another review round...
> 
> @jforissier, could you enable CI tests for this P-R? @raymo200915, it would be nice the CI includes a build test with `CFG_TRANSFER_LIST=y`. You can **~archive~ achieve** that by patching file _.github/workflows/ci.yaml_, see that file Git history for examples.
> 
> **(edited ;-)>**

Sorry for my last force-push...
Applied a minor fix for the stub function as @jenswi-linaro suggested.
Added CI build with CFG_TRANSFER_LIST and CFG_MAP_EXT_DT_SECURE.

etienne-lms reviewed (commented):
Thanks @raymo200915.
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for the series.

jenswi-linaro commented:
The latest fixups look good.

raymo200915 commented:
Squashed fixup commits and added Etienne's tag.

jforissier commented:
Thanks @raymo200915
